### ISSUE TITLE
Reverting to original list is problematic when running multiple builds at the same time

### ISSUE THREAD

**Issue by gdimartino-phunware** [richardszalay/spare-keys#3]
Hi, thank you so much for producing this awesome tool! It really made things easier for me. However, I am having a problem where multiple build jobs interfere with each other because of the way that the original list restored. Would it be possible to change the routine to just delete the added keychain from wherever it occurs in the list rather than simply restoring the original list? What is happening on our build server is that one job grabs the original list and adds it's own keychain to the top. Then another job grabs that new list and adds it's keychain to the top. When the first job is done, it restores it's version of the original list so that the second job now gets horsed up. At least that's what I think is happening :-)  Thank you for considering this issue.

**Comment by richardszalay** [comment]
Hi there, thanks for taking the time to log an issue. I'm glad to hear that you've found spare keys useful.

I had considered this scenario, but never got around to doing anything about it. The problem is that the current implementation sets the key chain as the _only_ keychain in the list in order to remove any chance of the build process attempting to fall back to a locked keychain.

Maybe we could form the behaviour behind a new `isolated` argument that would default to `true`. When `false`, it simply preprends the supplied/temp keychain to the list and then replaces it out again afterwards.

I don't have much time at the moment to make the change myself, but I'd happily accept a PR as long as you include some unit tests and add a note about the behaviour in the README.

**Comment by gdimartino-phunware** [comment]
Hi Richard,

Thanks for getting back to me! With respect to your comment about the current implementation setting the key chain as the only keychain in the list, my understanding was that it only did that if you passed in true for the clear_list parameter? I’m pretty sure it already allows the simple prepending of the supplied keychain if you set this parameter to false.

The problem is really in the way that it removes the supplied keychain when it is done. The way it does so now is by simply putting the previous list back (which can be bad if the list has changed since the original list was stored). I think the fix would be to grab the latest list and remove any occurrence of the supplied keychain. That way you are not blowing away any new keychains that have been added.

I will happily make the fix since you are OK with that, but please correct me if I’m wrong on any of this before I start.

Thanks!

## 

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com http://www.phunware.com/

**Comment by richardszalay** [comment]
Ah even better, I'd forgotten about `clear_list`. I'm happy for you to apply the change you described when `clear_list` is false.

**Comment by gdimartino-phunware** [comment]
Thank you! I am a Ruby newby, just so you know, so please feel free to instruct in any areas that might not be to your standards :-)

I will work on this as time permits, but hopefully over the next couple of days I will have something.

## 

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com http://www.phunware.com/

**Comment by richardszalay** [comment]
I'm sure it will be fine. spare-keys and fastlane-plugin-act were the first (and to this day, only) Ruby I've ever written.

**Comment by gdimartino-phunware** [comment]
Hi Richard,

I have something that I think will work, but I am having trouble figuring out how to fire your unit tests. I’m assume the code in the “spare_keys_spec.rb” file is the unit testing code (cause that’s what it looks like) and all I need to do is invoke the tests from the command line some how.

Also, it occurred to me that this is just a refactoring of existing code… there are no new features. So as long as the existing tests don’t fail I should be good, right?

Thanks,

## 

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com http://www.phunware.com/

**Comment by richardszalay** [comment]
I should probably add a "Contributing" section to the README, if only to remind me how all this work again. I believe you should be able to do the following:

```
bundle install # gem install bundler if you don't have it
rake
```

As for whether a new test is required - this is changing the existing behavior, so a new test should be created to cover expected behavior of `clear_list=false`. And yes, it should not break any existing `clear_list=true` tests.

Iet me know if have any issues.

**Comment by gdimartino-phunware** [comment]
Hi Richard,

After several interruptions, I managed to get something that I think will work. The new unit test passes and the old ones continue to pass as well.

I have checked in my changes to the branch I created locally called “issue-3”. In order to push that branch up to GitHub I think you need to make me a collaborator. Once you do, I can push my changes and create a pull request so that you can review the work and merge it into “master”.

## 

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com http://www.phunware.com/

**Comment by richardszalay** [comment]
Hi Gustavo,

That's great news. Actually you can create a Pull Request from within GitHub and I'll be able to merge in your branch from there.

Cheers,
Richard

**Comment by gdimartino-phunware** [comment]
Hi Richard,

I believe GitHub has to know about the branch before I can create a pull request, though, right? Currently, the branch only lives on my computer because I don't have permission to push it up. 

Gus

**Comment by richardszalay** [comment]
Hi Gus,

Now I understand. You'll need to:
1. Fork the repo to gdimartino-phunware/spare-keys
2. Change your remote by running `git remote set-url origin git@github.com:gdimartino-phunware/spare-keys.git`
3. Push your changes
4. Navigate to "https://github.com/gdimartino-phunware/spare-keys/tree/issue-3"
5. Click the "Create Pull Request" button

**Comment by gdimartino-phunware** [comment]
Thanks Richard, I'll try this in the morning and let you know how it goes!

G

**Comment by gdimartino-phunware** [comment]
Hi Richard,

I have an all day meeting today so I will give this a try this afternoon sometime.

Thanks!

## 

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com http://www.phunware.com/

**Comment by gdimartino-phunware** [comment]
Richard, so sorry for the delay, I am working on getting permission to fork your repo into our company account on Github.

--
Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com <http://www.phunware.com/>

**Comment by gdimartino-phunware** [comment]
Hi Richard,

I finally was able to take care of this. I wasn’t able to fork into my company’s account, so I just used my personal account.

So sorry this took so long. Looking forward to hearing from you!

Gustavo Di Martino
Software Engineer

Phunware, Inc.
The Software of Things™

(949) 300-5294 — Mobile

www.phunware.com <http://www.phunware.com/>

**Comment by richardszalay** [comment]
This PR has been merged and included in the 1.2.0 release
